google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageFileSystem.java

```diff
@@ -335,4 +335,45 @@ public String toString() {
       throw new AssertionError(e);
     }
   }
+
+  private static final class ConfigTuple {
+    private final CloudStorageConfiguration cloudStorageConfiguration;
+    @Nullable private final StorageOptions storageOptions;
+
+    public ConfigTuple(
+        CloudStorageConfiguration cloudStorageConfiguration,
+        @Nullable StorageOptions storageOptions) {
+      this.cloudStorageConfiguration =
+          requireNonNull(cloudStorageConfiguration, "cloudStorageConfiguration must be non null");
+      this.storageOptions = storageOptions;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (!(o instanceof ConfigTuple)) {
+        return false;
+      }
+      ConfigTuple that = (ConfigTuple) o;
+      return cloudStorageConfiguration.equals(that.cloudStorageConfiguration)
+          && Objects.equals(storageOptions, that.storageOptions);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(cloudStorageConfiguration, storageOptions);
+    }
```

Comment:
These overrides of `equals` and `hashCode` remind me of Brian's comment (2) in https://github.com/googleapis/java-storage-nio/issues/691#issuecomment-915605390 - is this now ok here because we're no longer looking up via `HashMap` (in which case is the override still necessary?) Or are we now waiting until after initialization before caching? Or is Brian's comment still potentially a problem?

Reply:
The `.equals` from that comment was comparing instances of `CloudStorageFileSystemProvider` which aren't immutable. Here we're holding on to `CloudStorageConfig` which is immutable via AutoValue and a `StorageOptions` which has a stable hashCode & equals which don't include transient state. So in this case, we're only keying off the config none of the running state.

Reply:
Great, thanks! Looks good to me!